crates/store/src/historical/mod.rs

```diff
@@ -0,0 +1,480 @@
+            }
+
+            // Update the leaf if it was modified in this overlay.
+            if let Some((key, value)) = overlay
+                .rev_set
+                .as_mutation_set()
+                .new_pairs()
+                .iter()
+                .find(|(k, _)| LeafIndex::from(**k) == leaf.index())
+            {
+                leaf = if *value == EMPTY_WORD {
+                    SmtLeaf::new_empty(leaf.index())
+                } else {
+                    SmtLeaf::new_single(*key, *value)
+                };
+            }
+        }
+
+        leaf
+    }
+
+    /// Builds a `SparseMerklePath` from the `path_nodes` array.
+    ///
+    /// The `empty_mask` is constructed by setting a bit for each empty node.
+    /// We iterate from depth 0 to `max_depth` in reverse order (high to low)
+    /// to build the nodes vector as expected by `SparseMerklePath`.
+    fn build_sparse_path(path_nodes: &[Option<Word>; SMT_DEPTH as usize]) -> (u64, Vec<Word>) {
+        let max_depth =
+            path_nodes.iter().rposition(std::option::Option::is_some).map_or(0, |d| d + 1);
+
+        // Start with all bits set (all empty), then clear bits for non-empty nodes.
+        let mut empty_mask = u64::MAX;
+        let nodes: Vec<Word> = (0..max_depth)
+            .rev()
+            .filter_map(|d| {
+                path_nodes[d].inspect(|_| {
+                    // Clear the bit at position `d` to indicate this node is present.
+                    empty_mask &= !(1u64 << d);
+                })
+            })
+            .collect();
+
+        (empty_mask, nodes)
+    }
+
+    /// Gets the account state at the latest block.
+    pub fn get(&self, account_id: AccountId) -> Word {
+        self.inner
+            .read()
+            .expect("RwLock poisoned: concurrent thread panicked while holding lock")
+            .latest
+            .get(account_id)
+    }
+
+    // PUBLIC MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Applies mutations and advances to the next block.
+    pub fn apply_mutations(
+        &self,
+        account_commitments: impl IntoIterator<Item = (AccountId, Word)>,
+    ) -> Result<(), HistoricalError> {
+        let mut inner = self
+            .inner
+            .write()
+            .expect("RwLock poisoned: concurrent thread panicked while holding lock");
+
+        let accounts = Vec::from_iter(account_commitments);
+
+        let mutations = inner.latest.compute_mutations(accounts)?;
+        let rev = inner.latest.apply_mutations_with_reversion(mutations)?;
+        let overlay = HistoricalOverlay::new(inner.block_number, rev);
+
+        inner.overlays.push_front(Arc::new(overlay));
+        inner.block_number = inner.block_number.child();
+        Self::drain_excess(&mut inner.overlays);
+
+        Ok(())
+    }
+
+    /// Returns the oldest block still in history.
+    pub fn oldest_block_num(&self) -> BlockNumber {
```

Comment:
AFAICT, this method does not mutate the state - so, it should probably not be in the "public mutators" section.

Similar comment for `contains_account_id_prefix()` method below.

Reply:
We do use it, it wasn't shown as used before since I wanted to hold off on using `AccountTreeWithHistory` with the remaining integration points.

```
crates/store/src/state.rs
870-        let new_account_id_prefix_is_unique = if account_commitment.is_empty() {
871:            Some(!inner.account_tree.contains_account_id_prefix(account_id.prefix()))
872-        } else {
873-            None
874-        };

```